serial abstraction layer (sal)

chronopoulos opened the pull request:
this PR introduces a "serial abstraction layer" called NewScaleSerial, which represents a platform-agnostic abstraction of the serial pipe to the USBInterface (M3-USB-3.1-EP). the NewScaleSerial class exposes a class method get_instances() which returns the instances of a serial port which match the VID and PID of the newscale device; on linux this is a pyserial "Serial" object, on windows it's a USBXpressDevice. instances of NewScaleSerial expose methods like get_port_name, get_serial_number, set_baudrate, set_timeout, write, and readLine which call the appropriate subroutines depending on the platform. a NewScaleSerial instance can be used to instantiate a USBInterface object, and then the rest of the library works in the same way.

examples/ have been updated, and i added back_and_forth_3d.py as a simple, repeatable (no random numbers) alternative to multistage_test.py. all examples have now been tested on both windows and linux.

chronopoulos committed:
add list_available_ports to USBXYZStage, USBInterface

chronopoulos committed:
list_available_ports: fix windows bug

chronopoulos committed:
add usbxpress.py

chronopoulos committed:
update README

chronopoulos committed:
add NewScaleSerial class

chronopoulos committed:
integrate into USBXYZStage

chronopoulos committed:
fix set_timeout for usbxpress

chronopoulos committed:
NewScaleSerial: fix timeout bug in readLine

chronopoulos committed:
add examples/moveBackAndForth_usb.py

chronopoulos committed:
examples/moveBackAndForth.py: clean imports, add YZ

chronopoulos committed:
rename examples/moveBackAndForth_usb.py

chronopoulos committed:
rename list_ports_test.py

chronopoulos committed:
clean up comments in NewScaleSerial, USBInterface

Commit:
examples/list_serial_devices.py: fix bug

Commit:
update examples/read_firmware.py

chronopoulos committed:
increment minor version

chronopoulos committed:
update constructor for USBXYZStage

chronopoulos committed:
update README.md

Poofjunior commented:
**License Stuff**

Does usbxpress.py have a license associated with it? I can't find a good source for where this file came from apart from the answer to [this forum question](https://community.silabs.com/s/question/0D51M00007xeQj9SAE/python-example-code-for-siusbxpdll?language=en_US). If there is no license, then we should probably reference their "[Master Software License Agreement](https://www.silabs.com/about-us/legal/master-software-license-agreement)."

Since the code is present as source code, it looks like this file might qualify as open source code, except that the file itself is not tagged as such. See:

> “Silicon Labs Open Source Code” means Software created by Silicon Labs and which is (a) delivered to Licensee in Source Code format, (b) is identified as open source code and (c) states that use of the software is subject to the terms of this Agreement.

In our case, I don't know about (b) and (c).

Is there a general license for code that appears in the forums?

Poofjunior commented on README.md:
```diff
@@ -29,23 +29,8 @@ stage = USBXYZStage('COM4')
 ## Usage
 Check the [examples directory](examples) for more usage examples.
 
```
This PR explicitly removes communication with M3 USB Devices through Windows as a COM port interface if the hardware was provisioned as such. I propose that we don't take away functionality but instead add to what's already there.

Generally, it looks like the M3 USB Devices can communicate through Windows through one of two drivers either (1) as a native usb device with a custom VID/PID besides the reserved ones or (2) as a COM port using a specific VID/PID.

I'd propose that we either (a) make two interfaces or (b) modify the M3USBInterface to communicate with these devices using the appropriate driver depending on parameters passed into the constructor. (I think choice b is better because it removes some of the responsibility of the user to know which classes to use.)

Here's an example of what that might look like:
````python
class USBInterface(HardwareInterface):
    def __init__(self, com_port: str = None, serial_number: str = None):
        # If the com_port is defined, communicate with the device as a COM port.
        # if the serial_number is defined, communicate with the device via the USBXpress.py library.
        # Error if both or neither parameter is specified.
        ...
````

Poofjunior replied:
Ok, second benefit to the approach above: you can hide all of the `NewScaleSerial.get_instances()` calls and stick them inside this class. I know you sacrifice the ability to "init the only instance" without knowing the serial number if you plug it in over USB. But that's an easy number to find in a benchtop setup, and a number we should know in advance in a more permanent setup.

Overall, I think device serial number is an ideal way to be able to instantiate this class since it is both fixed and unique across devices. Let's take advantage of what that level of granularity can offer us to hide the `get_instance` calls from the user.

Poofjunior reviewed (changes requested):
Overall, this looks really good, but we need to follow up on a couple things:
1. licenses on usbxpress.py
2. adding native usb functionality, rather than replacing the existing COM port functionality.
3. hiding the `get_instances` calls that the user doesn't need to do if they know the serial number.

Overall, really impressive work digging up Python bindings to the native USBXpress dll library. That affords us the ability to enumerate an M3 USB Device on the software side in a way that's both fixed and unique. I think the remaining part here would be to simplify instantiation from the user's point of view. I'm happy to take a stab at hiding the init stuff if you like, with the added consequence that I'll need to replicate your work on windows.

chronopoulos committed:
usbxpress.py: add copyright and license

chronopoulos committed:
NewScaleSerial: add support for PID=0xEA60 devices

Some New Scale devices may be provisioned with PID=0xEA60 (instead of
0xEA61), either from the factory or in-house.

On Windows, this invokes the Virtual Comport Driver (rather than
USBXpress). This change supports both PID's and drivers.

On Linux, both PID's invoke the same serial driver, and this change
merely updates the PID checking logic.

chronopoulos commented:
appreciate the review!

> 1. licenses on usbxpress.py

great point, i think this may actually fall under section 4.2.5 of the [MSLA](https://www.silabs.com/about-us/legal/master-software-license-agreement). i have added a copyright header and license reference as of 0dd801d, let me know what you think.

> 2. adding native usb functionality, rather than replacing the existing COM port functionality.

ok, while i don't think we should spend too much effort on supporting cases that New Scale doesn't support (like aftermarket PID modification), this is simple enough to keep that i agree with you. check out c4d61d9; i did something a little different than what you suggest but it achieves the same result: i just extended the NewScaleSerial.get_instances to include checks for the modified PID (0xea60), and then the rest is identical to the pyserial library as we use it in linux.

> 3. hiding the get_instances calls that the user doesn't need to do if they know the serial number.

i think i disagree with this suggestion, because asking "what new scale devices are connected right now" is a very common use case (i'm calling this in [parallax](https://github.com/AllenNeuralDynamics/parallax/blob/master/parallax/model.py#L64) to populate a drop-down menu with the serial numbers of available stages).

Poofjunior commented:
I can't speak to adding a license to code where there is no license in it already, but this might be the best we can do for now.

> check out https://github.com/AllenNeuralDynamics/python-newscale/commit/c4d61d93d2d3b282e676d799e7a4cd30cf6f2c5d

This looks great. That way we don't break linux functionality that we got for free.

> i think i disagree with this suggestion, because asking "what new scale devices are connected right now" is a very common use case (i'm calling this in [parallax](https://github.com/AllenNeuralDynamics/parallax/blob/master/parallax/model.py#L64) to populate a drop-down menu with the serial numbers of available stages).

Whoop I meant the following.

Let's keep what you have, i.e:
````python
devices = NewScaleSerial.get_instances()
````

But when it comes to instantiating a device, let's support both of these use cases:
````python
my_xyz_stage = M3USBInterface(devices[0]) 
# OR
my_xyz_stage = M3USBInterfaces(serial_num=46109)
````

That way you can support:
1. drop-down menu to list out devices (good for interactive prototyping)
2. a way of instantiating the device if you already know the serial number (good for permanent setups with a fixed configuration)

To do the latter, we just push the NewScaleSerial instantiation into the class USBInterface's `__init__`, rather than pass in an existing one as a parameter.

The only other thing I might suggest is considering making `get_instances` a standalone function to this module. Here's a [similar example from another package](https://github.com/python-ivi/python-usbtmc/blob/master/usbtmc/usbtmc.py#L123). Up to you on this one though!

Poofjunior commented on src/newscale/interfaces.py:
```diff
@@ -54,40 +62,116 @@ def read(self, address: str = None):
         raise NotImplementedError("To be implemented by the child class.")
 
 
+class NewScaleSerial():
```
Unless we are explicitly trying to support Python 2, let's stick with `class NewScaleSerial:` instead of `class NewScaleSerial():` since they both inherit from `object`.

[lengthy discussion on the matter](https://stackoverflow.com/a/45062077/3312269)

chronopoulos commented:
ok, i pushed a couple of changes: 0dd801d adds copyright and license, and c4d61d9 adds support for the provisioned devices.

`my_xyz_stage = M3USBInterface(devices[0]) `

are you aware that M3USBInterface is not used anywhere else in the library? it seems to only add setting/getting of baud rate, which i am currently handling at the NewscaleSerial level. i feel like we should either (a) remove these methods/class entirely or (b) move them into USBSerialInterface as a passthrough to NewscaleSerial. what do you think?

finally, i took a stab at implementing instantiation given a known serial number, but wasn't able to come up with anything i was happy with. maybe you could try and propose something?
